### ISSUE TITLE
Redeclaring a package variant should be at least a warning, probably an error

### ISSUE THREAD

**Issue by HadrienG2** [spack/spack#8922]
Currently, a package is allowed to declare a given variant twice, with the new definition silently overwriting the old one. I think it boils down to [this line in the code of variant()](https://github.com/spack/spack/blob/e2e8108dc4ca8210f1526ce68db59d633d53f66e/lib/spack/spack/directives.py#L494) not checking if a certain entry already exists in pkg.variants.

I think that spack should at least warn, and possibly bomb, on such variant re-declarations, as they probably stem from an accidental mistake rather than from a conscious design decision.

As an example of this happening in the wild, the "boost" package currently declares its "graph" variant twice, once setting the default value to True, and another time setting the default value to False. It took me a while to figure out this one, I'll submit a patch which removes the duplicate variant later today.

**Comment by adamjstewart** [comment]
I agree, although I can think of one situation where we may actually want this. If I have a package that uses another package as its base class we might want to be able to override its default variants. See the recent `mutil` PR for an example of what I'm talking about.

**Comment by becker33** [comment]
Overriding variants and composing conditional variants with multiple conditions are essential use cases for this, so we won't be able to forbid multiple variant definitions.

The new `spack info` output should help make it very clear when a variant is multiply defined.
